src/python/pants/engine/target.py

```diff
@@ -22,25 +24,55 @@ def __repr__(self) -> str:
 class PrimitiveField(Field, metaclass=ABCMeta):
     """A Field that does not need the engine in order to be hydrated.
 
-    This should be subclassed by the majority of fields.
+    This should be used by the majority of fields.
+
+    Subclasses must implement `hydrate()` to convert `self.raw_value` into `self.value`. This
+    hydration and/or validation happens eagerly in the constructor. If the hydration is
+    particularly expensive, use `AsyncField` instead to get the benefits of the engine's caching.
+
+    Subclasses should also override the type hints for `raw_value` and `value` to be more precise
+    than `Any`.
+
+    Example:
+
+        class ZipSafe(PrimitiveField):
+            alias: ClassVar = "zip_safe"
+            raw_value: Optional[bool]
+            value: bool
+
+            def hydrate(self) -> bool:
+                if self.raw_value is None:
+                    return True
+                return self.raw_value
     """
 
-    def __str__(self) -> str:
-        return f"{self.alias}={self.value}"
+    value: Any
+
+    @final
```

Comment:
This is a really neat feature from [PEP 591](https://www.python.org/dev/peps/pep-0591/) that tells MyPy to check that no subclasses are overriding the method.